x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/AttributeMapTests.java

```diff
@@ -249,4 +251,83 @@ public void testCopy() {
 
         assertThat(m, is(copy));
     }
+
+    public void testEmptyMapIsImmutable() {
+        var empty = AttributeMap.emptyAttributeMap();
+        var ex = expectThrows(UnsupportedOperationException.class, () -> empty.add(a("one"), new Object()));
```

Comment:
question/suggestion: doesn't this potentially lead to lots of specialcasing wherever client code needs to add something to an `AttributeMap`? Technically, I'd need to check whether I have the empty attribute map whenever I call `AttributeMap::add`, no? Also, seems like conceptually the empty map has a different contract than a general attribute map.

IMHO it'd be better to have `emptyAttributeMap` return a freshly-made empty map, then all attribute maps would be consistently mutable.

Reply:
It follow the convention of Java collections.  The client can create an empty collection which is mutable - however for the common case where an immutable empty collection is used, there's emptyAttributeMap similar to Collections.emptyMap()/Set()/empty().
The whole promise of the method is to return a constant - the same old empty map; which needs to be immutable since otherwise it gets modified and thus all the other places where it is being used now contain non-empty maps.

Reply:
Ah, thanks for clarifying - I was unaware of the convention.